notebooks/work-in-progress/ferc1-transform-refactor/ferc1_transform_refactor.py

```diff
@@ -0,0 +1,429 @@
+    repl=r"\1_mwh",
+)
+"""Parameters for converting column units from kWh to MWh."""
+
+BTU_TO_MMBTU = dict(
+    multiplier=1e-6,
+    pattern=r"(.*)_btu(.*)$",
+    repl=r"\1_mmbtu\2",
+)
+"""Parameters for converting column units from BTU to MMBTU."""
+
+FUEL_TYPES = {
+    "categories": {
+        "coal": {"black", "xenoforming", "dinosaur", "goo"},
+        "oil": {"olive", "sunflower", "fish", "astroglide"},
+        "gas": {"elon musk", "unicorn farts", "cow farts"},
+    }
+}
+
+TRANSFORM_PARAMS = {
+    "test_table_one": {
+        "rename_columns": {
+            "columns": {
+                "netgen": "net_generation_kwh",
+                "heat_content": "fuel_btu_per_unit",
+                "units": "fuel_units",
+                "fuel_type": "fuel_type",
+            }
+        },
+        "convert_units": {
+            "net_generation_kwh": KWH_TO_MWH,
+            "fuel_btu_per_unit": BTU_TO_MMBTU,
+        },
+        "categorize_strings": {
+            "fuel_type": FUEL_TYPES,
+        },
+    },
+    "test_table_two": {
+        "rename_columns": {
+            "columns": {
+                "netgen": "net_generation_kwh",
+                "heat_content": "fuel_btu_per_unit",
+                "units": "fuel_units",
+                "fuel_type": "fuel_type",
+            }
+        },
+        "convert_units": {
+            "net_generation_kwh": KWH_TO_MWH,
+            "fuel_btu_per_unit": BTU_TO_MMBTU,
+        },
+        "categorize_strings": {
+            "fuel_type": FUEL_TYPES,
+        },
+    },
+}
+
+
+################################################################################
+# Pydanic Transformation Parameter Models
+################################################################################
+class TransformParams(BaseModel):
+    """An immutable base model for transformation parameters."""
+
+    class Config:
+        """Prevent parameters from changing part way through."""
+
+        allow_mutation = False
+
+
+class MultiColumnTransformParams(TransformParams):
+    """Transform params that apply to several columns in a table.
+
+    The keys are column names, and the values must all be the same type of
+    :class:`TransformParams` object, since MultiColumnTransformParams are used by
+    :class:`MultiColumnTransformFn` callables.
+
+    Individual subclasses are dynamically generated for each multi-column transformation
+    specified within a :class:`TableTransformParams` object.
+
+    """
+
+    @root_validator
+    def single_param_type(cls, params):  # noqa: N805
+        """Check that all TransformParams in the dictionary are of the same type."""
+        param_types = {type(params[col]) for col in params}
+        if len(param_types) > 1:
+            raise ValueError(
+                "Found multiple parameter types in multi-column transform params: "
+                f"{param_types}"
+            )
+        return params
+
+
+class RenameColumns(TransformParams):
```

Comment:
should this be a `MultiColumnTransformParams`? just confused how these relate bc rename will definitely happen to many columns

Reply:
RenameColumns applies at the table level. It does operate on many columns, but it is NOT a composite of column level transforms.